botocore/utils.py

```diff
@@ -260,70 +265,125 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
+        )
+
+    def _contains_all_credential_fields(self, credentials):
+        for field in self._REQUIRED_CREDENTIAL_FIELDS:
+            if field not in credentials:
+                logger.debug(
+                    'Retrieved credentials is missing required field: %s',
+                    field)
+                return False
+        return True
+
+    def _is_non_ok_response(self, response):
+        if response.status_code != 200:
+            self._log_imds_response(response, 'non-200', log_body=True)
+            return True
+        return False
+
+    def _is_empty(self, response):
+        if not response.content:
+            self._log_imds_response(response, 'no body', log_body=True)
+            return True
+        return False
+
+    def _is_invalid_json(self, response):
         try:
-            r = self._get_request(url, timeout, num_attempts)
-            if r.content:
-                fields = r.content.decode('utf-8').split('\n')
-                for field in fields:
-                    if field.endswith('/'):
-                        data[field[0:-1]] = self.retrieve_iam_role_credentials(
-                            url + field, timeout, num_attempts)
-                    else:
-                        val = self._get_request(
-                            url + field,
-                            timeout=timeout,
-                            num_attempts=num_attempts,
-                        ).content.decode('utf-8')
-                        if val[0] == '{':
-                            val = json.loads(val)
-                        data[field] = val
-            else:
-                logger.debug("Metadata service returned non 200 status code "
-                             "of %s for url: %s, content body: %s",
-                             r.status_code, url, r.content)
-        except _RetriesExceededError:
-            logger.debug("Max number of attempts exceeded (%s) when "
-                         "attempting to retrieve data from metadata service.",
-                         num_attempts)
-        # We sort for stable ordering. In practice, this should only consist
-        # of one role, but may need revisiting if this expands in the future.
-        final_data = {}
-        for role_name in sorted(data):
-            final_data = {
-                'role_name': role_name,
-                'access_key': data[role_name]['AccessKeyId'],
-                'secret_key': data[role_name]['SecretAccessKey'],
-                'token': data[role_name]['Token'],
-                'expiry_time': data[role_name]['Expiration'],
-            }
-        return final_data
+            json.loads(response.text)
+            return False
+        except ValueError:
+            self._log_imds_response(response, 'invalid json')
+            return True
+
+    def _log_imds_response(self, response, reason_to_log, log_body=False):
+        statement = (
+            "Metadata service returned " + reason_to_log + " response "
```

Comment:
How come `reason_to_log` is the only one not being formatted in at log time?

Reply:
That was a refactoring artifact. I had `reason_to_log` optional at some point. I can update it.